google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerCloudMonitoringExporter.java

```diff
@@ -0,0 +1,246 @@
+      spannerTimeSeries =
+          SpannerCloudMonitoringExporterUtils.convertToSpannerTimeSeries(spannerMetricData);
+    } catch (Throwable e) {
+      logger.log(
+          Level.WARNING,
+          "Failed to convert spanner metric data to cloud monitoring timeseries.",
+          e);
+      return CompletableResultCode.ofFailure();
+    }
+
+    ProjectName projectName = ProjectName.of(spannerProjectId);
+
+    ApiFuture<List<Empty>> futureList = exportTimeSeriesInBatch(projectName, spannerTimeSeries);
+
+    CompletableResultCode spannerExportCode = new CompletableResultCode();
+    ApiFutures.addCallback(
+        futureList,
+        new ApiFutureCallback<List<Empty>>() {
+          @Override
+          public void onFailure(Throwable throwable) {
+            if (spannerExportFailureLogged.compareAndSet(false, true)) {
+              String msg = "createServiceTimeSeries request failed for spanner metrics.";
+              if (throwable instanceof PermissionDeniedException) {
+                // TODO: Add the link of public documentation when available in the log message.
+                msg +=
+                    String.format(
+                        " Need monitoring metric writer permission on project=%s.",
+                        projectName.getProject());
+              }
+              logger.log(Level.WARNING, msg, throwable);
+            }
+            spannerExportCode.fail();
+          }
+
+          @Override
+          public void onSuccess(List<Empty> empty) {
+            // When an export succeeded reset the export failure flag to false so if there's a
+            // transient failure it'll be logged.
+            spannerExportFailureLogged.set(false);
+            spannerExportCode.succeed();
+          }
+        },
+        MoreExecutors.directExecutor());
+
+    return spannerExportCode;
+  }
+
+  private ApiFuture<List<Empty>> exportTimeSeriesInBatch(
+      ProjectName projectName, List<TimeSeries> timeSeries) {
+    List<ApiFuture<Empty>> batchResults = new ArrayList<>();
+
+    for (List<TimeSeries> batch : Iterables.partition(timeSeries, EXPORT_BATCH_SIZE_LIMIT)) {
+      CreateTimeSeriesRequest req =
+          CreateTimeSeriesRequest.newBuilder()
+              .setName(projectName.toString())
+              .addAllTimeSeries(batch)
+              .build();
+      ApiFuture<Empty> f = this.client.createServiceTimeSeriesCallable().futureCall(req);
+      batchResults.add(f);
+    }
+
+    return ApiFutures.allAsList(batchResults);
+  }
+
+  @Override
+  public CompletableResultCode flush() {
+    if (lastExportCode != null) {
+      return lastExportCode;
+    }
+    return CompletableResultCode.ofSuccess();
+  }
+
+  @Override
+  public CompletableResultCode shutdown() {
+    if (!isShutdown.compareAndSet(false, true)) {
+      logger.log(Level.WARNING, "shutdown is called multiple times");
+      return CompletableResultCode.ofSuccess();
+    }
+    CompletableResultCode flushResult = flush();
```

Comment:
The SDK usually do this for you before calling this method. Check out our implementation https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/blob/11c85b9e3809a60a6348982ee9ea38d28cbe49db/exporters/metrics/src/main/java/com/google/cloud/opentelemetry/metric/InternalMetricExporter.java#L271-L276

Reply:
There is not much difference in what is done in user facing exporter and what is done here except some extra checks. 

Here as well we are calling `metricServiceClient.shutdown()` , its just we are having extra try catch block here to make sure that any error in these exporter does not propagate to customer. 

Also we are explicitly calling `flush` in the `shutdown` .

Reply:
OK after digging a bit more, I agree flush() makes sense.

One question though, if an exception is thrown in the `whenComplete`, does it actually propagate to the user? I would imagine the error is surfaced to the caller of `OpenTelemetrySdk.shutdown()`

Reply:
I think its ok to surround in try-catch to prevent any error leaking to customer.

However, `flush()` does not actually perform any actual export for metrics, what is the reason behind the explicit call in shutdown?

Reply:
Removed the flush() call from shutdown as flush is not doing anything